### PR TITLE
cluster: Support TiKV-CDC component

### PR DESCRIPTION
**Pull request by pingyu**
Issue Number: #1999

Signed-off-by: pingyu <yuping@pingcap.com>

### What problem does this PR solve? 

#1999 

### What is changed and how it works?

TiUP cluster supports the [TiKV-CDC](https://github.com/tikv/migration/tree/main/cdc) component.

### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)

```sh
➜  tiup-cluster deploy up v6.2.0 up.yaml
...
...
Cluster `up` deployed successfully, you can start it with command: `tiup-cluster start up --init`
➜  tiup-cluster start up -y
Starting cluster up...
...
...
Started cluster `up` successfully
➜  tiup-cluster display up
Cluster type:       tidb
Cluster name:       up
Cluster version:    v6.2.0
Deploy user:        root
SSH type:           builtin
...
...
Total nodes: 14
➜ tikv-cdc cli --pd 10.2.x.x:2379 capture list
[
  {
    "id": "3e2ede2a-13ad-4ded-8ac6-1c5f315dfd5e",
    "is-owner": false,
    "address": "10.2.x.x:8600"
  },
  {
    "id": "7b68a84e-63d5-4fd2-89a6-90aa15a7069c",
    "is-owner": false,
    "address": "10.2.x.x:8600"
  },
  {
    "id": "a05ae66d-6dcf-4c28-b0d2-0cab3a221dfc",
    "is-owner": true,
    "address": "10.2.x.x:8600"
  }
]
➜
```

Code changes

 - Has exported variable/fields change
   - spec.TiKVCDCSpec, scripts.TiKVCDCScript, api.TiKVCDCOpenAPIClient
   - spec.ServerConfigs.TiKVCDC, spec.Specification.TiKVCDCServers, instance.ComponentTiKVCDC

Side effects

 - Increased code complexity

Related changes

 - Need to update the documentation
   - TBD

Release notes:

```release-note
TiUP cluster supports the `TiKV-CDC` component.
```

### CODE REVIEW

**Comment by pingyu** [comment]
@nexustar PTAL, thanks~

**Review by haojinming — APPROVED** [review]
LGTM~

**Comment by zeminzhou** [comment]
LGTM~

**Review by zeminzhou — APPROVED** [review]

File: pkg/cluster/spec/bindversion.go [hunk 1]
```diff
@@ -28,7 +28,8 @@ func TiDBComponentVersion(comp, version string) string {
 		ComponentPushwaygate,
 		ComponentCheckCollector,
 		ComponentSpark,
-		ComponentTiSpark:
+		ComponentTiSpark,
+		ComponentTiKVCDC: // TiKV-CDC use individual version.
```
Comment (AstroProfundis) [inline pkg/cluster/spec/bindversion.go:32]: Is there any compatibility requirement or the user can always use the latest version of TiKV-CDC regardless of the TiDB version?

File: pkg/cluster/spec/spec.go [hunk 2]
```diff
@@ -112,6 +112,7 @@ type (
 		Pump           map[string]interface{} `yaml:"pump"`
 		Drainer        map[string]interface{} `yaml:"drainer"`
 		CDC            map[string]interface{} `yaml:"cdc"`
+		TiKVCDC        map[string]interface{} `yaml:"tikv-cdc"`
```
Comment (AstroProfundis) [inline pkg/cluster/spec/spec.go:115]: Better use `tikv_cdc` to keep the same naming format with other components.

File: pkg/cluster/spec/spec.go [hunk 3]
```diff
@@ -127,6 +128,7 @@ type (
 		PumpServers      []*PumpSpec          `yaml:"pump_servers,omitempty"`
 		Drainers         []*DrainerSpec       `yaml:"drainer_servers,omitempty"`
 		CDCServers       []*CDCSpec           `yaml:"cdc_servers,omitempty"`
+		TiKVCDCServers   []*TiKVCDCSpec       `yaml:"tikv-cdc_servers,omitempty"`
```
Comment (AstroProfundis) [inline pkg/cluster/spec/spec.go:131]: ditto, `tikv_cdc_servers`

File: tests/tiup-cluster/script/cmd_subtest.sh [hunk 4]
```diff
@@ -5,15 +5,48 @@ set -eu
...
+                ;;
+            --tls)
+                test_tls=true
+                shift
+                ;;
+            --native-ssh)
+                native_ssh=true
+                shift
+                ;;
+            --staging)
```
Comment (AstroProfundis) [inline tests/tiup-cluster/script/cmd_subtest.sh:35]: This staging mirror is not used anywhere in this PR

File: tests/tiup-cluster/script/tikv_cdc.sh [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+
+set -eu
+
+function tikv_cdc_test() {
+	mkdir -p ~/.tiup/bin/
+
+	version="nightly"
+	topo_name="tikv_cdc"
+	test_tls=false
+	mirror="https://tiup-mirrors.pingcap.com/"
```
Comment (AstroProfundis) [inline tests/tiup-cluster/script/tikv_cdc.sh:11]: `tiup mirror set --reset` will do the same as `tiup mirror set https://tiup-mirrors.pingcap.com`

File: tests/tiup-cluster/script/tikv_cdc.sh [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+		--tikv-cdc-patch)
+			tikv_cdc_patch="$2"
+			shift
+			shift
+			;;
+		esac
+	done
+
+	# set mirror
+	tiup mirror set $mirror
```
Comment (AstroProfundis) [inline tests/tiup-cluster/script/tikv_cdc.sh:43]: as the staging mirror is not used, i think it's not necessary to build `tiup` and use it to manually set the mirror.

File: pkg/cluster/spec/tikv_cdc.go [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+)
+
+// TiKVCDCSpec represents the TiKVCDC topology specification in topology.yaml
+type TiKVCDCSpec struct {
+	Host            string                 `yaml:"host"`
+	SSHPort         int                    `yaml:"ssh_port,omitempty" validate:"ssh_port:editable"`
+	Imported        bool                   `yaml:"imported,omitempty"`
+	Patched         bool                   `yaml:"patched,omitempty"`
+	IgnoreExporter  bool                   `yaml:"ignore_exporter,omitempty"`
+	Port            int                    `yaml:"port" default:"8600"`
```
Comment (AstroProfundis) [inline pkg/cluster/spec/tikv_cdc.go:37]: The port should also be added to `portTypes` list in `pkg/cluster/spec/validate.go` at around L619, other wise the port conflict check is not working for it.

File: pkg/cluster/spec/spec.go (hunk 2 above)
Comment (pingyu) [reply pkg/cluster/spec/spec.go:115]: The reason why I chose `tikv-cdc` is that:
1. I refer to this [PR](https://github.com/pingcap/tiup/pull/2017) that using `tidb-dashboard` & `tidb-dashboard_servers`.
2. The binary is named as `tikv-cdc`, so using `tikv_cdc` will be error-prone, e.g. `run_tikv-cdc.sh.tpl` will has both `tikv-cdc` & `tikv_cdc`.

Or how about using `kvcdc` ? Just as what we add to playground (#2000) ?

File: pkg/cluster/spec/bindversion.go (hunk 1 above)
Comment (pingyu) [reply pkg/cluster/spec/bindversion.go:32]: No requirement. User can always use the latest version by now, and we will maintain the compatibility.

File: pkg/cluster/spec/spec.go (hunk 2 above)
Comment (AstroProfundis) [reply pkg/cluster/spec/spec.go:115]: I think #2017 should also change to use identical name formatting, I'm absolutely sure that mixing `-` and `_` in the same key name will cause chaos to the end users.

`kvcdc` would be a good choice IMO.

Comment (pingyu) [reply pkg/cluster/spec/spec.go:115]: Okay~ I will change to `kvcdc`.

File: pkg/cluster/spec/tikv_cdc.go (hunk 7 above)
Comment (pingyu) [reply pkg/cluster/spec/tikv_cdc.go:37]: I found that the `Port` is already in the list for [`portTypes`](https://github.com/pingcap/tiup/blob/3671d82636e29dd71f86ba6fb3cf6eaec141dda9/pkg/cluster/spec/validate.go#L620) where `portConflictsDetect` get port value from. So the port conflict check should work.

And I add a test case to verify port conflict check in integrated test for tikv-cdc. PTAL~

File: tests/tiup-cluster/script/tikv_cdc.sh (hunk 6 above)
Comment (pingyu) [reply tests/tiup-cluster/script/tikv_cdc.sh]: `tiup` is also used for installing `tikv-cdc` component for test case of `tiup cluster patch`. I remove the `tiup mirror` here, but keep the `tiup` target & function.

File: pkg/cluster/spec/spec.go (hunk 2 above)
Comment (pingyu) [reply pkg/cluster/spec/spec.go]: Done. PTAL~

**Comment by pingyu** [comment]
@AstroProfundis Thanks for your review ! All comments are addressed or replied, PTAL~

**Review by nexustar — APPROVED** [review]

**Comment by pingyu** [comment]
Added cluster version check. PTAL, thanks~

**Pull request merged**